data_mgr.py

```diff
@@ -0,0 +1,299 @@
+    # Use already established client
+    try:
+        pb_msg = await client.async_request(method)
+    except ClientTimeout as exc:
+        return (id, MSG_TIMEOUT)
+    except ClientError as exc:
+        return (id, None)
+    else:
+        ws_data = METHOD_MAP[method]()
+        ws_data.ParseFromString(pb_msg)
+    return (id, ws_data)
+
+
+class DataManager(object):
+
+    def __init__(self, ws_mgr):
+        self.ws_mgr = ws_mgr
+        self.data = {}
+
+    # **
+    # Data syncing
+    # **
+    async def entire_workflow_update(self, ids=None):
+        if ids is None:
+            ids = []
+        data = {}
+        method = 'pb_entire_workflow'
+        ws_args = (
+            (id, info['req_client'], method)
+            for id, info in self.ws_mgr.workflows.items())
+        # print(self.ws_mgr.workflows)
+        items = await gen.multi(
+            get_workflow_data(*args) for args in ws_args
+            if not ids or args[0] in ids)
+        for id, result in items:
+            if result is not None and result != MSG_TIMEOUT:
+                data[id] = result
+        # atomic update
+        if not ids:
+            self.data = data
+        else:
+            for id, flow_msg in data.items():
+                self.data[id] = flow_msg
+
+    # **
+    # Data access filters
+    # **
+    def get_workflow_msgs(self, args):
+        """Return list of workflows."""
+        ids = [self.parse_workflow_id(id) for id in args['ids']]
+        exids = [self.parse_workflow_id(id) for id in args['exids']]
+        result = []
+        for flow_msg in self.data.values():
+            if self._workflow_filter(flow_msg, ids, exids):
+                result.append(flow_msg)
+        return result
+
+    def get_nodes_all(self, node_type, args):
+        """Return list of nodes (defns/proxies/jobs)."""
+        w_args = {
+            'ids': args['workflows'],
+            'exids': args['exworkflows'],
+        }
+        nodes = [
+            n for k in self.get_workflow_msgs(w_args)
+            for n in getattr(k, node_type)]
+        ids = [self.parse_node_id(id, node_type) for id in args['ids']]
+        exids = [self.parse_node_id(id, node_type) for id in args['exids']]
+        result = []
+        for node in nodes:
+            if self._node_filter(node, ids, exids, args):
+                result.append(node)
+        return result
+
+    def get_nodes_id(self, node_type, args):
+        """Return GraphQL data for given id."""
+        nat_ids = args.get('native_ids', [])
+        w_ids = []
+        for nat_id in nat_ids:
+            oname, wname, nid = nat_id.split('/', 2)
+            w_ids.append(f"{oname}/{wname}")
+        if node_type == 'proxy_nodes':
+            nodes = (
+                n for w_id in set(w_ids)
+                for n in (
+                    list(getattr(self.data[w_id], 'task_proxies', []))
+                    + list(getattr(self.data[w_id], 'family_proxies', []))))
```

Comment:
There is a bit of a logical pattern `list(getattr(self.data[w_id], node_type, []))` in here and below. Perhaps you can introduce a function or method to improve readability?

Reply:
No. Protobuf message lists don't support `+` ... Happy to implement something else if you have any suggestions..